test/integration/puppetlabs/services/jruby/jruby_pool_int_test.clj

```diff
@@ -283,3 +292,84 @@
+    (let [services [jruby/jruby-puppet-pooled-service profiler/puppet-profiler-service
+                    handler-service/request-handler-service ps-config/puppet-server-config-service
+                    jetty9/jetty9-service]
+          config (-> (jruby-testutils/jruby-puppet-tk-config
+                      (jruby-testutils/jruby-puppet-config {:max-active-instances 2}))
+                     (assoc-in [:webserver :port] 8081))
+          app (tk/boot-services-with-config services config)
+          cert (ssl-utils/pem->cert
+                (str test-resources-dir "/localhost-cert.pem"))
+          jruby-service (tk-app/get-service app :JRubyPuppetService)
+          jruby-instance (jruby-protocol/borrow-instance jruby-service :i-want-this-instance)
+          handler-service (tk-app/get-service app :RequestHandlerService)
+          request { :uri "/puppet/v3/environments", :params {}, :headers {},
+                   :request-method :GET, :body "", :ssl-client-cert cert, :content-type ""}
+          ping-environment #(->> request (handler-core/wrap-params-for-jruby) (handler/handle-request handler-service))
+          stop-complete? (future (tk-app/stop app))]
+        (let [start (System/currentTimeMillis)]
+          (while (and
+                  (< (- (System/currentTimeMillis) start) 10000)
+                  (not= 503 (:status (ping-environment))))
+            (Thread/yield))
+          (is (= 503 (:status (ping-environment))))
+          (jruby-protocol/return-instance jruby-service jruby-instance :i-want-this-instance)
+          @stop-complete?
+          (is (= 503 (:status (ping-environment))))))))
+
+(deftest ^:integration test-503-when-jruby-is-first-to-shutdown
+  (testing "During a shutdown requests result in 503 http responses"
+    (bootstrap/with-puppetserver-running
+     app
+     {:jruby-puppet {:max-active-instances 2}}
+     (let [jruby-service (tk-app/get-service app :JRubyPuppetService)
+           context (tk-services/service-context jruby-service)
+           jruby-instance (jruby-protocol/borrow-instance jruby-service :i-want-this-instance)
+           stop-complete? (future (tk-services/stop jruby-service context))
+           ping-environment #(http-client/get "https://localhost:8140/puppet/v3/environments"
+                                              bootstrap/request-options)]
+       (let [start (System/currentTimeMillis)]
+         (while (and
+                 (< (- (System/currentTimeMillis) start) 10000)
+                 (not= 503 (:status (ping-environment))))
+           (Thread/yield)))
+       (is (= 503 (:status (ping-environment))))
+       (jruby-protocol/return-instance jruby-service jruby-instance :i-want-this-instance)
+       @stop-complete?
+       (let [app-context (tk-app/app-context app)]
+         (swap! app-context assoc :JRubyPuppetService {})
+         (tk-internal/run-lifecycle-fn! app-context tk-services/init "init" :JRubyPuppetService jruby-service)
```

Comment:
Because we reached in and stopped the jruby-service earlier in the test, we need to reinitialize it here for the tk stop to actually succeed.